src/app/Library/CrudPanel/Traits/Create.php

```diff
@@ -236,4 +237,42 @@ private function getRelationDataFromFormData($data)
 
         return $relationData;
     }
+
+    /**
+     * Transform the data before saving into the database. This is kind of an alias to Laravel Mutators but defined on Controler when defining the field.
+     * We basically call this "mutator" before saving the data into the database but after data is already validated.
+     *
+     * @param  array  $data  The form data.
+     * @return array The processed data.
+     */
+    protected function handleBeforeSavingCallback($data)
+    {
+        $fields = array_filter($this->fields(), function ($field) use ($data) {
+            return in_array($field['name'], array_keys($data)) && isset($field['beforeSaving']);
+        });
+
+        if (empty($fields)) {
+            return $data;
+        }
+        // cicle all the fields that have the beforeSaving callback
+        foreach ($fields as $field_name => $field_definition) {
+            if (! is_array($field_definition['name'])) {
+                $data[$field_name] = $this->runBeforeSavingCallback($data[$field_name], $field_name, $field_definition);
+            } else {
+                foreach ($field_definition['name'] as $f_name) {
+                    $data[$f_name] = $this->runBeforeSavingCallback($data[$f_name], $f_name, $field_definition);
+                }
+            }
+        }
+
+        return $data;
+    }
+
+    protected function runBeforeSavingCallback($current, $field_name, $field)
+    {
+        $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->{$field_name} : null;
+        if (is_callable($field['beforeSaving'])) {
+            return $field['beforeSaving']($current, $previous_values, $field_name);
```

Comment:
I'm wondering what parameters the callback should have. Probably only those that the dev can't get himself, right? The previous value... they can get that if they do `CRUD::entry()` inside the callback - correct?

What will the dev need to do _most of the time_ in this callback? Would they need the previous value often - enough for us to make it dead-simple to get it, by having it as a parameter?

Similarly, can't they get the old value using `old($field['name'])`? Would that be inconvenient?

Because callback parameters... once they're there, we can't change them without a breaking change. So we gotta think as simple as possible here. 

In the end I was thinking a simple but inclusive method signature would be `function($value, $field)` whereby `$field` is the entire field definition array - that way they'd have access to `$field['name']` but also any other attributes like the model or something else... I don't know.

What do you think @pxpm ? What does the dev need to do here most of the time? Let's simplify.

Reply:
I am now refactoring this. 

I agree with most of your reasoning, but while refactoring I noticed something odd. Passing the `field name` is not necessary I think, because developer knows the field name while defining the callback ? 

Bare with me, you are defining the field:
```php
$this->crud->addField([
    'name' => 'my_field_name',
    'beforeSaving' => function($values) {
         // why would dev need here to have the field or the field name in the parameters if he is defining the field ? He knows the field name and the field parameters. 
         $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->my_field_name : null;
    }
]);
```

So I don't think having `field` or `field_name` would improve this. 

Either we `pre-load` the previous values and have `function($current, $previous)` or simply have `function($current)` and developer can get the previous if needed.

But I am not sure if we should settle only in `current` and `previous`. Maybe `$data` (the request validated that is going to be saved in the database). should be passed too ? 

```php
$this->crud->addField([
    'name' => 'my_field_name',
    'beforeSaving' => function($data) {
         $current_values = $data['my_field_name'];
         $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->my_field_name : null;

        // do whatever you need to do. 
    }
]);

$this->crud->addField([
    'name' => 'my_field_name',
    'beforeSaving' => function($data, $previous) {
         $current_values = $data['my_field_name'];
        // do whatever you need to do. 
    }
]);

$this->crud->addField([
    'name' => 'my_field_name',
    'beforeSaving' => function($data, $current, $previous) {
        // do whatever you need to do. 
    }
]);
```

**I like the last option more.** I am thinking in a crud where I have 3/4 or more fields with save callback, If I have to get common information in all of them it would get very repetitive to watch. As example:

```php
$this->crud->addFields([
    [ 'name' => 'my_field_name',
    'beforeSaving' => function($data) {
         $current_values = $data['my_field_name'];
         $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->my_field_name : null;

        // do whatever you need to do. 
    }],
    [ 'name' => 'other_field',
    'beforeSaving' => function($data) {
         $current_values = $data['other_field'];
         $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->other_field: null;

        // do whatever you need to do. 
    }],
    [ 'name' => 'third_field',
    'beforeSaving' => function($data) {
         $current_values = $data['third_field'];
         $previous_values = $this->getCurrentEntry() ? $this->getCurrentEntry()->third_field: null;

        // do whatever you need to do. 
    }],
]);

```

Let me know what you think, to sum up, I would settle with `function($data,$current,$previous)`, even if not used a small task like `if(Str::startsWith('a', $previous') { return strtoupper($current) }` could get very verbose if developer need to fetch everything in all their callbacks.